### PR TITLE
Boosted DeepTauID v2.0

### PR DESCRIPTION
**Pull request by mbluj**
#### PR description:

This PR adds DeepTauID dedicated to boostedTaus (aka taus in boosted final states).
This new tauID is run at miniAOD step and added to boostedTau collections at miniAOD and nanoAOD. It can be also run on-the-fly as part on nanoAOD step with existing miniAOD samples (for validation purposes) with a new era modifier `run3_nanoAOD_pre142X`  introduced in #46066.

Expected changes:
* three new tauIDs in the `slimmedTausBoosted` collection in miniAOD (with negligible impact on miniAOD size);
* three new tauIDs (three new columns) in the `boostedTau` table in nanoAOD and relaxed boostedTau pre-selection (pt threshold 40->25GeV, boostedDeepTauID added to selection criteria OR'ed with previous tauIDs) - increase of size of the `boostedTau` table by ~70% and of size of nanoAOD by ~0.5% (tested with Run3 ttbar MC).

This PR depends on DeepTauID model files added in https://github.com/cms-data/RecoTauTag-TrainingFiles/pull/14.

Most recent presentation on the BoostedDeepTauID: https://indico.cern.ch/event/1444695/#2-deepboostedtau-id-for-hh-bbt

#### PR validation:

Validated with wf. 13034.21 (TTbar_14TeV+2024PU_ProdLike) 
and custom nanoAOD wf with 2022 miniAOD and `run3_nanoAOD_pre142X` modifier with cmsDriver command: 
```
cmsDriver.py myNANO2022 -s NANO --mc --conditions auto:phase1_2022_realistic --era Run3,run3_nanoAOD_pre142X --eventcontent NANOAODSIM --datatier NANOAODSIM -n 1000 --no_exec --filein file:/eos/cms/store/group/phys_tau/kandroso/miniAOD/Run3_2022/TTtoLNu2Q.root
```

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Backport not planned

### CODE REVIEW

**Comment by mbluj** [comment]
FYI @gparida @Ksavva1021 @pritampalit

**Comment by jfernan2** [comment]
enable nano

**Comment by jfernan2** [comment]
please test

**Comment by mbluj** [comment]
It should be tested with https://github.com/cms-data/RecoTauTag-TrainingFiles/pull/14 otherwise model files are not available.

**Comment by jfernan2** [comment]
please test with  https://github.com/cms-data/RecoTauTag-TrainingFiles/pull/14

**Comment by mbluj** [comment]
This failure is expected as with the current setup it is expected that new BoostedDeepTauID is present in miniAOD unless it is added to nanoAOD workflow by employing one of the following modifiers: `run3_nanoAOD_pre142X`, `run3_nanoAOD_124`, `run2_nanoAOD_106Xv2`. As far as I understand a fix of the observed issues with matrix tests is implemented in https://github.com/cms-sw/cmssw/pull/46066. @hqucms could you confirm, please?

**Comment by mbluj** [comment]
Last commit adds forgotten plots with the boostedDeepTauID to Nano's DQM

**Comment by hqucms** [comment]
test_parameters:
 - pull_requests = https://github.com/cms-data/RecoTauTag-TrainingFiles/pull/14
 - enable = profiling,nano

**Comment by hqucms** [comment]
please test

**Comment by jfernan2** [comment]
please test
Nano tests show increases in rate of processed events of up to 20%, let's trigger the tests again to decouple from merge commits which were also included on top of IB

File: PhysicsTools/NanoAOD/python/nano_cff.py [hunk 1]
```diff
@@ -281,7 +281,12 @@ def nanoAOD_customizeCommon(process):
         idsToAdd = cms.vstring()
     )
     run2_nanoAOD_106Xv2.toModify(
-        nanoAOD_boostedTau_switch, idsToAdd = ["mvaIso", "mvaIsoNewDM", "mvaIsoDR0p3", "againstEle"]
+        nanoAOD_boostedTau_switch, idsToAdd = ["mvaIso", "mvaIsoNewDM", "mvaIsoDR0p3", "againstEle", "boostedDeepTauRunIIv2p0"]
+    ).toModify(
+        process, lambda p : nanoAOD_addBoostedTauIds(p, nanoAOD_boostedTau_switch.idsToAdd.value())
+    )
+    (run3_nanoAOD_124 | run3_nanoAOD_pre142X).toModify(
```
Comment (hqucms) [inline PhysicsTools/NanoAOD/python/nano_cff.py:288]: Please remove `run3_nanoAOD_124` as we plan to clean up that modifier.

File: RecoTauTag/RecoTau/python/tools/runTauIdMVA.py [hunk 2]
```diff
@@ -1314,11 +1345,12 @@ def getDeepTauVersion(self, file_name):
         """returns the DeepTau year, version, subversion. File name should contain a version label with data takig year \
         (2011-2, 2015-8), version number (vX) and subversion (pX), e.g. 2017v0p6, in general the following format: \
         {year}v{version}p{subversion}"""
-        version_search = re.search('(20[1,2][125678])v([0-9]+)(p[0-9]+|)[\\._]', file_name)
+        version_search = re.search('(20[1,2][125678]|RunII)v([0-9]+)(p[0-9]+|)[\\._]', file_name)
         if not version_search:
             raise RuntimeError('File "{}" has an invalid name pattern, should be in the format "{year}v{version}p{subversion}". \
                                 Unable to extract version number.'.format(file_name))
         year = version_search.group(1)
+        if year == 'RunII': year = 20161718
```
Comment (hqucms) [inline RecoTauTag/RecoTau/python/tools/runTauIdMVA.py:1353]: I suppose this ID is trained with Run2 samples, so has it been tested to see if the performance still holds for Run3 samples?

**Review by hqucms — COMMENTED** [review]
Looking at the [CPU profiling result](https://cmssdt.cern.ch/circles/web/piechart.php?local=false&dataset=CMSSW_14_2_X_2024-09-26-2300%2Fel8_amd64_gcc12%2F12634.21%2FPR-70ad07%2F41782%2Fstep4_cpu.resources&resource=time_thread&colours=default&groups=packages&filter=CMSSW_14_2_X_2024-09-26-2300&data_name=profiling&show_labels=true&threshold=0), I see that this boosted tau ID takes almost as much time as DeepTau v2.5, while naively I would expected it to be much faster as there should be fewer boosted taus. Did I miss anything? Or could the inference time be further optimized for this boosted tau ID?
 - deepTau2018v2p5ForMini    30.5 ms
 - boostedDeepTau20161718v2p0BoostedForMini    27.5 ms

File: RecoTauTag/RecoTau/python/tools/runTauIdMVA.py (hunk 2 above)
Comment (mbluj) [reply RecoTauTag/RecoTau/python/tools/runTauIdMVA.py:1353]: Yes, it was trained on mix of MC samples from all 2016-18 Run2 years to cope with limited statistics (I think mainly signal samples). It was checked that performance of the BoostedDeepTauID tagger is better that old BDT-based taggers with Run-3 data, however it is worse than for Run-2 data. It seems that degradation is similar as for the regular DeepTauID v2.5. Probably a dedicated training with Run3 data would be better, but it is not on the place so current one is the best available.

**Comment by mbluj** [comment]
> Looking at the [CPU profiling result](https://cmssdt.cern.ch/circles/web/piechart.php?local=false&dataset=CMSSW_14_2_X_2024-09-26-2300%2Fel8_amd64_gcc12%2F12634.21%2FPR-70ad07%2F41782%2Fstep4_cpu.resources&resource=time_thread&colours=default&groups=packages&filter=CMSSW_14_2_X_2024-09-26-2300&data_name=profiling&show_labels=true&threshold=0), I see that this boosted tau ID takes almost as much time as DeepTau v2.5, while naively I would expected it to be much faster as there should be fewer boosted taus. Did I miss anything? Or could the inference time be further optimized for this boosted tau ID?
> 
> * deepTau2018v2p5ForMini    30.5 ms
> * boostedDeepTau20161718v2p0BoostedForMini    27.5 ms

BoostedDeepTauID shares architecture with regular DeepTauID (v2.5) so inference time per tau candidate of the two should be similar. In the semileptonic ttbar miniAOD sample number of boostedTau candidates (mainly fakes) is equal to about 60% of regular tau candidates (also mainly fakes) so I would expect that inference time of boostedDeepTauID be also roughly 60% of inference time of regular one. But it could be that in samples used for profiling the ratio is different.
The only simple way to improve it in a short timescale is to tighten selection of boosted-tau candidates, but with impact on analyses difficult to asses.

**Comment by hqucms** [comment]
> > Looking at the [CPU profiling result](https://cmssdt.cern.ch/circles/web/piechart.php?local=false&dataset=CMSSW_14_2_X_2024-09-26-2300%2Fel8_amd64_gcc12%2F12634.21%2FPR-70ad07%2F41782%2Fstep4_cpu.resources&resource=time_thread&colours=default&groups=packages&filter=CMSSW_14_2_X_2024-09-26-2300&data_name=profiling&show_labels=true&threshold=0), I see that this boosted tau ID takes almost as much time as DeepTau v2.5, while naively I would expected it to be much faster as there should be fewer boosted taus. Did I miss anything? Or could the inference time be further optimized for this boosted tau ID?
> > 
> > * deepTau2018v2p5ForMini    30.5 ms
> > * boostedDeepTau20161718v2p0BoostedForMini    27.5 ms
> 
> BoostedDeepTauID shares architecture with regular DeepTauID (v2.5) so inference time per tau candidate of the two should be similar. In the semileptonic ttbar miniAOD sample number of boostedTau candidates (mainly fakes) is equal to about 60% of regular tau candidates (also mainly fakes) so I would expect that inference time of boostedDeepTauID be also roughly 60% of inference time of regular one. But it could be that in samples used for profiling the ratio is different. The only simple way to improve it in a short timescale is to tighten selection of boosted-tau candidates, but with impact on analyses difficult to asses.

I see -- it's probably ok for now, but in the long run it would be worth to switch to a more efficient model architecture.

**Comment by hqucms** [comment]
please test

**Comment by hqucms** [comment]
+1

**Comment by makortel** [comment]
Just to "test drive" the recently added monitoring of maximum allocated memory at any time, I see this PR increases maximum allocated memory of the step2 of many NANO workflows by ~35-40 MB ([link](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisonsNANO/CMSSW_14_2_X_2024-09-30-1100+70ad07/64808/maxmem-comparison/)). Is that expected?

**Comment by mbluj** [comment]
This PR adds DeepTauID to to boosted taus in the nanoAOD step in "legacy" workflows while it is added to the miniAOD step in the "future" workflows.The DeepTau is a big model and known to allocate quite some memory. However, I do not understand why in some (usually MC workflows) the effect is different (smaller).

**Comment by jfernan2** [comment]
@mbluj can the differences in the workflows be explained by the presence/absence of taus? I see some are leptonic decays while others not:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-70ad07/41782/runTheMatrixNANO-results/

**Comment by mbluj** [comment]
> @mbluj can the differences in the workflows be explained by the presence/absence of taus? I see some are leptonic decays while others inclusive: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-70ad07/41782/runTheMatrixNANO-results/

Yes, it can be the reason. Leptons also give rise to tau candidates, but occupancy of deepTau grid is different for letons and jets, also multiplicity of tau candidates is probably different in different samples.

**Comment by jfernan2** [comment]
+1

**Comment by antoniovilela** [comment]
@mbluj Could you please look at the conflicts?

**Comment by mbluj** [comment]
> @mbluj Could you please look at the conflicts?

Rebase to master should solve them. I'll do it, run local tests and then push here.

**Comment by mbluj** [comment]
Conflicts solved.

**Comment by mandrenguyen** [comment]
@jfernan2 Please also sign https://github.com/cms-data/RecoTauTag-TrainingFiles/pull/14

**Comment by mandrenguyen** [comment]
please test

**Comment by smuzaffar** [comment]
please test

**Comment by mbluj** [comment]
Tests started 18h ago and not results yet, should the tests be restarted?

**Comment by hqucms** [comment]
please abort

**Comment by hqucms** [comment]
please test

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**